src/coreclr/nativeaot/Runtime/unix/PalRedhawkInline.h

```diff
@@ -17,84 +17,81 @@ FORCEINLINE void PalInterlockedOperationBarrier()
     //
     // It is possible for a load following the code sequence above to be reordered to occur prior to the store above due to the
     // release barrier, this is substantiated by https://github.com/dotnet/coreclr/pull/17508. Interlocked operations in the PAL
-    // require the load to occur after the store. This memory barrier should be used following a call to a __sync* function to
-    // prevent that reordering. Code generated for arm32 includes a 'dmb' after 'cbnz', so no issue there at the moment.
-    __sync_synchronize();
+    // require the load to occur after the store. This memory barrier should be used following a call to an atomic operation to
+    // prevent that reordering. Code generated for ARM32 includes a 'dmb' after 'cbnz', so no issue there at the moment.
+    __atomic_thread_fence(__ATOMIC_SEQ_CST);
 #endif
 }
 
 FORCEINLINE int32_t PalInterlockedIncrement(_Inout_ int32_t volatile *pDst)
 {
-    int32_t result = __sync_add_and_fetch(pDst, 1);
+    int32_t result = __atomic_add_fetch(pDst, 1, __ATOMIC_SEQ_CST);
     PalInterlockedOperationBarrier();
     return result;
 }
 
 FORCEINLINE int32_t PalInterlockedDecrement(_Inout_ int32_t volatile *pDst)
 {
-    int32_t result = __sync_sub_and_fetch(pDst, 1);
+    int32_t result = __atomic_sub_fetch(pDst, 1, __ATOMIC_SEQ_CST);
     PalInterlockedOperationBarrier();
     return result;
 }
 
 FORCEINLINE uint32_t PalInterlockedOr(_Inout_ uint32_t volatile *pDst, uint32_t iValue)
 {
-    int32_t result = __sync_or_and_fetch(pDst, iValue);
+    uint32_t result = __atomic_or_fetch(pDst, iValue, __ATOMIC_SEQ_CST);
     PalInterlockedOperationBarrier();
     return result;
 }
 
 FORCEINLINE uint32_t PalInterlockedAnd(_Inout_ uint32_t volatile *pDst, uint32_t iValue)
 {
-    int32_t result = __sync_and_and_fetch(pDst, iValue);
+    uint32_t result = __atomic_and_fetch(pDst, iValue, __ATOMIC_SEQ_CST);
     PalInterlockedOperationBarrier();
     return result;
 }
 
 FORCEINLINE int32_t PalInterlockedExchange(_Inout_ int32_t volatile *pDst, int32_t iValue)
 {
-#ifdef __clang__
-    int32_t result =__sync_swap(pDst, iValue);
-#else
     int32_t result =__atomic_exchange_n(pDst, iValue, __ATOMIC_ACQ_REL);
-#endif
     PalInterlockedOperationBarrier();
     return result;
 }
 
 FORCEINLINE int64_t PalInterlockedExchange64(_Inout_ int64_t volatile *pDst, int64_t iValue)
 {
-#ifdef __clang__
-    int32_t result =__sync_swap(pDst, iValue);
-#else
     int32_t result =__atomic_exchange_n(pDst, iValue, __ATOMIC_ACQ_REL);
-#endif
     PalInterlockedOperationBarrier();
     return result;
 }
 
 FORCEINLINE int32_t PalInterlockedCompareExchange(_Inout_ int32_t volatile *pDst, int32_t iValue, int32_t iComparand)
 {
-    int32_t result = __sync_val_compare_and_swap(pDst, iComparand, iValue);
+    int32_t expected = iComparand;
+    __atomic_compare_exchange_n(pDst, &expected, iValue, false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
     PalInterlockedOperationBarrier();
```

Comment:
One of the reason we insert an explicit memory barrier here is to achieve SEQ_CST semantics - is it still needed with the new instructions? cc @VSadov

Reply:
The way it's defined in pal.h, we can align with it here. There, the cast is only applicable to RV64 and LA64, arm64 was conditions on !__clang__ as a precaution for gcc in https://github.com/am11/runtime/commit/2820a1e775e29c5695505b6e391c73a581093643, which doesn't seem necessary either, so I've just cleaned it up.

@filipnavara might have some thoughts if we can align pal.h and palrh?